CONTRIBUTING.md

```diff
@@ -514,6 +515,49 @@ this API we will not break anyone, because they weren't able to use it. The file
 `allowed-breaking-changes.txt` in the root of the repo is an exclusion file that
 can be used in these cases.
 
+### Feature Flags
+
+Sometimes we want to introduce new breaking behavior because we believe this is
+the correct default behavior for the CDK. The problem of course is that breaking
+changes are only allowed in major versions and those are rare.
+
+To address this need, we have a feature flags pattern/mechanism. It allows us to
+introduce new breaking behavior which is disabled by default (so existing
+projects will not be affected) but enabled automatically for new projects
+created through `cdk init`.
+
+The pattern is simple:
+
+1. Define a new const under
+   [cx-api/lib/features.ts](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/cx-api/lib/features.ts)
+   with the name of the context key that **enables** this new feature (for
+   example, `ENABLE_STACK_NAME_DUPLICATES`). The context key should be in the
+   form `module.Type:feature` (e.g. `@aws-cdk/core:enableStackNameDuplicates`).
+2. Use `node.tryGetContext(cxapi.ENABLE_XXX)` to check if this feature is enabled
+   in your code. If it is not defined, revert to the legacy behavior.
+3. Add your feature flag to
+   [cx-api/lib/future.ts](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/cx-api/lib/future.ts).
+   This map is inserted to generated `cdk.json` files for new projects created
+   through `cdk init`.
+4. In your PR title (which goes into CHANGELOG), add a `(under feature flag)` suffix. e.g:
+
+    ```
+    fix(core): impossible to use the same physical stack name for two stacks (under feature flag)
+    ```
+5. Under `BREAKING CHANGES` in your commit message describe this new behavior:
+
+    ```
+    BREAKING CHANGE: template file names for new projects created through "cdk init" 
```

Comment:
If it's behind a feature flag, it's not a breaking change. I would suggest calling this out differently, so it doesn't alarm users and get conservative in their upgrades.

Reply:
Good point. Do we need some new wording for these cases? Something like "Breaking change behind feature flag"?

Reply:
@nija-at if new projects include the feature flag in their context, there's still a behavior change if `cdk init` was being called. It's still unexpected behavior and not predictable to the client who may have automated those calls.

I agree calling it out in some meaningful way is ideal, since it's a special type of breaking change

Reply:
+1 on @shivlaks's response. It is a breaking change, but only for new projects. I think that adding a `(under feature flag)` suffix is good enough. Also technically, our changelog generator only supports a single type of "BREAKING CHANGES" section, so I wouldn't want to over-complicate this.

Reply:
How about we change this to be a prefix (somewhat like @skinny85 suggests:

```
BREAKING CHANGE: (under feature flag) bla bla bla
```

What do you think?

Reply:
added

Reply:
Apologies to reopen this. I don't think this message is sufficient either. 

We have not published any information about 'feature flags' (i.e., after this PR is merged) to consumers of the cdk; and even if we do, we're making an assumption that they've heard of the (rarely used) concept of feature flags and know what such a comment means.

We should reserve the words `BREAKING CHANGE` to actually breaking changes. I would rather use terms like `Behaviour change for new projects` or something along those lines. Alternately we can use the words `Under a feature flag:` and hyperlink to a document that explains feature flags.

Reply:
We currently don’t have a way to render anything other than breaking changes in our changelog. Let’s just add a section to our docs that describe feature flags and explain what it means that it’s a “breaking change under a feature flag”. I agree that this is not a terminology we used and That’s a good call out. I’ll add to the implementation plan.

Reply:
I still think "behind feature flag" sounds better than "under feature flag" 🤡

Reply:
Fixed